sys/mips/gt64120.c

```diff
@@ -158,19 +159,46 @@ static void gt_pci_unmask_irq(intr_event_t *ie) {
   gt_pci_set_icus(gtpci);
 }
 
-static void gt_pci_intr_setup(device_t *pcib, unsigned irq,
-                              intr_handler_t *handler) {
-  assert(pcib->parent->driver == &gt_pci_bus.driver);
+/* clang-format off */
+static const char *gt_pci_intr_name[16] = {
+  [0] = "timer",
+  [1] = "kbd",        /* kbd controller (keyboard) */
+  [2] = "pic-slave",  /* PIC cascade */
+  [3] = "uart(1)",    /* COM 2 */
+  [4] = "uart(0)",    /* COM 1 */
+  [5] = "unused(0)",
+  [6] = "floppy",     /* floppy */
+  [7] = "parallel",   /* centronics */
+  [8] = "rtc",        /* RTC */
+  [9] = "i2c",        /* I2C */
+  [10] = "unused(1)",
+  [11] = "unused(2)",
+  [12] = "mouse",     /* kbd controller (mouse) */
+  [13] = "unused(3)",
+  [14] = "ide(0)",    /* IDE primary */
+  [15] = "ide(1)",    /* IDE secondary */
+};
+/* clang-format on */
 
-  gt_pci_state_t *gtpci = pcib->parent->state;
-  intr_event_t *event = &gtpci->intr_event[irq];
-  intr_event_add_handler(event, handler);
+static void gt_pci_intr_setup(device_t *dev, resource_t *r, ih_filter_t *filter,
+                              ih_service_t *service, void *arg,
+                              const char *name) {
+  assert(dev->parent->driver == &gt_pci_bus.driver);
+  gt_pci_state_t *gtpci = dev->parent->state;
+  int irq = r->r_start;
+
+  if (gtpci->intr_event[irq] == NULL)
+    gtpci->intr_event[irq] = intr_event_create(
+      gtpci, irq, gt_pci_mask_irq, gt_pci_unmask_irq, gt_pci_intr_name[irq]);
+
+  r->r_handler =
+    intr_event_add_handler(gtpci->intr_event[irq], filter, service, arg, name);
 }
 
-static void gt_pci_intr_teardown(device_t *pcib, intr_handler_t *handler) {
+static void gt_pci_intr_teardown(device_t *pcib, resource_t *irq) {
   assert(pcib->parent->driver == &gt_pci_bus.driver);
 
-  intr_event_remove_handler(handler);
+  intr_event_remove_handler(irq->r_handler);
```

Comment:
We are leaking an interrupt handler here.

Reply:
Done.